engine/src/vault/view.rs

```diff
@@ -9,11 +9,39 @@ use crate::vault::{
     },
 };
 
+use runtime::GuardedVec;
 use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, convert::Infallible, fmt::Debug};
+use thiserror::Error as DeriveError;
 
-use std::collections::HashMap;
+use super::crypto_box::DecryptError;
 
-use runtime::GuardedVec;
+#[derive(DeriveError, Debug)]
+pub enum VaultError<TProvErr: Debug, TProcErr: Debug = Infallible> {
+    #[error("Vault `{0:?}` does not exist.")]
+    VaultNotFound(VaultId),
+
+    #[error("Record Error: `{0:?}`")]
+    Record(#[from] RecordError<TProvErr>),
+
+    #[error("Procedure Error `{0:?}`")]
+    Procedure(TProcErr),
+}
+
+#[derive(DeriveError, Debug)]
+pub enum RecordError<TProvErr: Debug> {
+    #[error("Provider Error: `{0:?}`")]
+    Provider(TProvErr),
+
+    #[error("Found Invalid Transaction")]
+    CorruptedContent,
+
+    #[error("Invalid Key provided")]
```

Comment:
Error messages should be lowercase (we're also doing this wrong :sweat_smile:). I'm fairly sure this is mentioned somewhere in error handling best practices, but can't find it right now. Instead, quoting _Rust for Rustaceans_:

> In general, your implementation of Display should give a one-line description of what went wrong that can easily be folded into other error messages. The display format should be **lowercase** and without trailing punctuation so that it fits nicely into other, larger error reports.